File: src/core/contract-ai/types/definition.ts
/**
 * Contract AI - Layer 1: Definition Types (CO)
 * 
 * Definiuje STRUKTURĘ aplikacji - encje, eventy, workflow, API.
 * 
 * @version 2.2.0
 * @see todo/14-reclapp-llm-code-generation-spec.md
 */

// ============================================================================
// FIELD TYPES
// ============================================================================

/**
 * Podstawowe typy pól obsługiwane przez Contract AI
 */
export type BasicFieldType = 
  | 'String'
  | 'Int'
  | 'Float'
  | 'Boolean'
  | 'UUID'
  | 'DateTime';

/**
 * Rozszerzone typy pól z walidacją
 */
export type ExtendedFieldType =
  | 'Email'
  | 'URL'
  | 'Phone'
  | 'Money'
  | 'JSON'
  | 'Text';

/**
 * Typy relacji między encjami
 */
export type RelationType =
  | 'OneToOne'
  | 'OneToMany'
  | 'ManyToOne'
  | 'ManyToMany';

/**
 * Wszystkie obsługiwane typy pól
 */
export type FieldType = BasicFieldType | ExtendedFieldType | string;

// ============================================================================
// FIELD ANNOTATIONS
// ============================================================================

/**
 * Adnotacje pola definiujące walidację i zachowanie
 * 
 * @example
 * ```typescript
 * const emailAnnotations: FieldAnnotations = {
 *   required: true,
 *   unique: true,
 *   pattern: '^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\\.[a-zA-Z]{2,}$'
 * };
 * ```
 */
export interface FieldAnnotations {
  /** Pole jest wymagane */
  required?: boolean;
  /** Wartość musi być unikalna */
  unique?: boolean;
  /** Wartość generowana automatycznie */
  generated?: boolean;
  /** Wartość domyślna */
  default?: any;
  /** Minimalna wartość (dla liczb) lub długość (dla stringów) */
  min?: number;
  /** Maksymalna wartość (dla liczb) lub długość (dla stringów) */
  max?: number;
  /** Wzorzec regex do walidacji */
  pattern?: string;
  /** Dozwolone wartości (enum) */
  enum?: string[];
  /** Nazwa relacji do innej encji */
  relation?: string;
  /** Opis pola */
  description?: string;
}

// ============================================================================
// FIELD DEFINITION
// ============================================================================

/**
 * Definicja pola encji
 * 
 * @example
 * ```typescript
 * const emailField: FieldDefinition = {
 *   name: 'email',
 *   type: 'Email',
 *   annotations: {
 *     required: true,
 *     unique: true
 *   }
 * };
 * ```
 */
export interface FieldDefinition {
  /** Nazwa pola */
  name: string;
  /** Typ pola */
  type: FieldType;
  /** Adnotacje pola */
  annotations?: FieldAnnotations;
}

// ============================================================================
// RELATION DEFINITION
// ============================================================================

/**
 * Definicja relacji między encjami
 * 
 * @example
 * ```typescript
 * const contactCompanyRelation: RelationDefinition = {
 *   name: 'company',
 *   type: 'ManyToOne',
 *   target: 'Company',
 *   foreignKey: 'companyId',
 *   onDelete: 'SET NULL'
 * };
 * ```
 */
export interface RelationDefinition {
  /** Nazwa relacji */
  name: string;
  /** Typ relacji */
  type: RelationType;
  /** Docelowa encja */
  target: string;
  /** Pole klucza obcego */
  foreignKey?: string;
  /** Akcja przy usunięciu */
  onDelete?: 'CASCADE' | 'SET NULL' | 'RESTRICT' | 'NO ACTION';
  /** Akcja przy aktualizacji */
  onUpdate?: 'CASCADE' | 'SET NULL' | 'RESTRICT' | 'NO ACTION';
}

// ============================================================================
// ENTITY DEFINITION
// ============================================================================

/**
 * Definicja encji domenowej
 * 
 * @example
 * ```typescript
 * const contactEntity: EntityDefinition = {
 *   name: 'Contact',
 *   description: 'Customer contact information',
 *   fields: [
 *     { name: 'id', type: 'UUID', annotations: { generated: true } },
 *     { name: 'email', type: 'Email', annotations: { required: true, unique: true } },
 *     { name: 'firstName', type: 'String', annotations: { required: true } }
 *   ],
 *   relations: [
 *     { name: 'company', type: 'ManyToOne', target: 'Company' }
 *   ]
 * };
 * ```
 */
export interface EntityDefinition {
  /** Nazwa encji (PascalCase) */
  name: string;
  /** Opis encji */
  description?: string;
  /** Pola encji */
  fields: FieldDefinition[];
  /** Relacje do innych encji */
  relations?: RelationDefinition[];
  /** Indeksy bazy danych */
  indexes?: IndexDefinition[];
  /** Tagi dla kategoryzacji */
  tags?: string[];
}

/**
 * Definicja indeksu bazy danych
 */
export interface IndexDefinition {
  /** Nazwa indeksu */
  name: string;
  /** Pola w indeksie */
  fields: string[];
  /** Czy indeks jest unikalny */
  unique?: boolean;
}

// ============================================================================
// EVENT DEFINITION
// ============================================================================

/**
 * Definicja eventu domenowego
 * 
 * @example
 * ```typescript
 * const contactCreatedEvent: EventDefinition = {
 *   name: 'ContactCreated',
 *   description: 'Emitted when a new contact is created',
 *   fields: [
 *     { name: 'contactId', type: 'UUID', annotations: { required: true } },
 *     { name: 'email', type: 'Email', annotations: { required: true } }
 *   ],
 *   triggers: ['api.contact.create', 'import.contact']
 * };
 * ```
 */
export interface EventDefinition {
  /** Nazwa eventu (PascalCase) */
  name: string;
  /** Opis eventu */
  description?: string;
  /** Pola eventu */
  fields: FieldDefinition[];
  /** Wyzwalacze eventu */
  triggers?: string[];
  /** Encja źródłowa */
  sourceEntity?: string;
}

// ============================================================================
// WORKFLOW DEFINITION
// ============================================================================

/**
 * Typy kroków workflow
 */
export type WorkflowStepType =
  | 'action'
  | 'condition'
  | 'parallel'
  | 'loop'
  | 'wait'
  | 'notify'
  | 'transform'
  | 'validate';

/**
 * Definicja kroku workflow
 */
export interface WorkflowStepDefinition {
  /** ID kroku */
  id: string;
  /** Typ kroku */
  type: WorkflowStepType;
  /** Nazwa kroku */
  name: string;
  /** Opis kroku */
  description?: string;
  /** Konfiguracja kroku */
  config?: Record<string, any>;
  /** Następny krok (dla success) */
  next?: string;
  /** Krok przy błędzie */
  onError?: string;
  /** Warunek wykonania */
  condition?: string;
}

/**
 * Definicja workflow
 * 
 * @example
 * ```typescript
 * const onboardingWorkflow: WorkflowDefinition = {
 *   name: 'CustomerOnboarding',
 *   description: 'New customer onboarding process',
 *   trigger: 'ContactCreated',
 *   steps: [
 *     { id: 'verify', type: 'action', name: 'Verify Email', next: 'welcome' },
 *     { id: 'welcome', type: 'notify', name: 'Send Welcome Email' }
 *   ]
 * };
 * ```
 */
export interface WorkflowDefinition {
  /** Nazwa workflow */
  name: string;
  /** Opis workflow */
  description?: string;
  /** Wersja workflow */
  version?: string;
  /** Event wyzwalający */
  trigger?: string;
  /** Harmonogram (cron) */
  schedule?: string;
  /** Kroki workflow */
  steps: WorkflowStepDefinition[];
}

// ============================================================================
// API DEFINITION
// ============================================================================

/**
 * Operacje CRUD
 */
export type ApiOperation = 'list' | 'get' | 'create' | 'update' | 'delete';

/**
 * Definicja zasobu API
 */
export interface ApiResourceDefinition {
  /** Nazwa zasobu (kebab-case, plural) */
  name: string;
  /** Encja źródłowa */
  entity: string;
  /** Dostępne operacje */
  operations: ApiOperation[];
  /** Dodatkowe endpointy */
  customEndpoints?: ApiEndpointDefinition[];
}

/**
 * Definicja customowego endpointu
 */
export interface ApiEndpointDefinition {
  /** Metoda HTTP */
  method: 'GET' | 'POST' | 'PUT' | 'PATCH' | 'DELETE';
  /** Ścieżka (względna do zasobu) */
  path: string;
  /** Opis endpointu */
  description?: string;
  /** Schemat request body */
  requestBody?: Record<string, any>;
  /** Schemat response */
  response?: Record<string, any>;
}

/**
 * Definicja API
 * 
 * @example
 * ```typescript
 * const crmApi: ApiDefinition = {
 *   version: 'v1',
 *   prefix: '/api/v1',
 *   resources: [
 *     { name: 'contacts', entity: 'Contact', operations: ['list', 'get', 'create', 'update', 'delete'] },
 *     { name: 'companies', entity: 'Company', operations: ['list', 'get', 'create', 'update'] }
 *   ]
 * };
 * ```
 */
export interface ApiDefinition {
  /** Wersja API */
  version: string;
  /** Prefix URL */
  prefix: string;
  /** Zasoby API */
  resources: ApiResourceDefinition[];
  /** Autentykacja */
  authentication?: {
    type: 'jwt' | 'api-key' | 'oauth2' | 'none';
    config?: Record<string, any>;
  };
}

// ============================================================================
// APP DEFINITION
// ============================================================================

/**
 * Definicja aplikacji
 * 
 * @example
 * ```typescript
 * const crmApp: AppDefinition = {
 *   name: 'CRM System',
 *   version: '1.0.0',
 *   description: 'Customer Relationship Management System'
 * };
 * ```
 */
export interface AppDefinition {
  /** Nazwa aplikacji */
  name: string;
  /** Wersja aplikacji */
  version: string;
  /** Opis aplikacji */
  description?: string;
  /** Autor */
  author?: string;
  /** Licencja */
  license?: string;
}

// ============================================================================
// DEFINITION LAYER
// ============================================================================

/**
 * Layer 1: Definition - definiuje CO ma być zaimplementowane
 * 
 * @example
 * ```typescript
 * const crmDefinition: DefinitionLayer = {
 *   app: {
 *     name: 'CRM System',
 *     version: '1.0.0',
 *     description: 'Customer Relationship Management'
 *   },
 *   entities: [
 *     { name: 'Contact', fields: [...] },
 *     { name: 'Company', fields: [...] }
 *   ],
 *   api: {
 *     version: 'v1',
 *     prefix: '/api/v1',
 *     resources: [...]
 *   }
 * };
 * ```
 */
export interface DefinitionLayer {
  /** Definicja aplikacji */
  app: AppDefinition;
  /** Encje domenowe */
  entities: EntityDefinition[];
  /** Eventy domenowe */
  events?: EventDefinition[];
  /** Workflow */
  workflows?: WorkflowDefinition[];
  /** Definicja API */
  api?: ApiDefinition;
}

File: src/python/reclapp/models/definition.py
"""
Contract AI - Layer 1: Definition Types (CO)

Defines the STRUCTURE of the application - entities, events, workflow, API.

Mirrors: src/core/contract-ai/types/definition.ts
"""

from enum import Enum
from typing import Any, Literal, Optional, Union

from pydantic import BaseModel, Field


# ============================================================================
# FIELD TYPES
# ============================================================================

class BasicFieldType(str, Enum):
    """Basic field types supported by Contract AI"""
    STRING = "String"
    INT = "Int"
    FLOAT = "Float"
    BOOLEAN = "Boolean"
    UUID = "UUID"
    DATETIME = "DateTime"


class ExtendedFieldType(str, Enum):
    """Extended field types with validation"""
    EMAIL = "Email"
    URL = "URL"
    PHONE = "Phone"
    MONEY = "Money"
    JSON = "JSON"
    TEXT = "Text"


# All supported field types
FieldType = Union[BasicFieldType, ExtendedFieldType, str]


class RelationType(str, Enum):
    """Relation types between entities"""
    ONE_TO_ONE = "OneToOne"
    ONE_TO_MANY = "OneToMany"
    MANY_TO_ONE = "ManyToOne"
    MANY_TO_MANY = "ManyToMany"


# ============================================================================
# FIELD ANNOTATIONS
# ============================================================================

class FieldAnnotations(BaseModel):
    """
    Field annotations defining validation and behavior
    
    Example:
        annotations = FieldAnnotations(
            required=True,
            unique=True,
            pattern=r'^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$'
        )
    """
    required: Optional[bool] = None
    unique: Optional[bool] = None
    generated: Optional[bool] = None
    default: Optional[Any] = None
    min: Optional[int] = None
    max: Optional[int] = None
    pattern: Optional[str] = None
    enum: Optional[list[str]] = None
    relation: Optional[str] = None
    description: Optional[str] = None


# ============================================================================
# FIELD DEFINITION
# ============================================================================

class FieldDefinition(BaseModel):
    """
    Entity field definition
    
    Example:
        field = FieldDefinition(
            name="email",
            type="Email",
            annotations=FieldAnnotations(required=True, unique=True)
        )
    """
    name: str
    type: str  # FieldType
    annotations: Optional[FieldAnnotations] = None


# ============================================================================
# RELATION DEFINITION
# ============================================================================

class RelationDefinition(BaseModel):
    """
    Relation definition between entities
    
    Example:
        relation = RelationDefinition(
            name="company",
            type=RelationType.MANY_TO_ONE,
            target="Company",
            foreignKey="companyId",
            onDelete="SET NULL"
        )
    """
    name: str
    type: RelationType
    target: str
    foreignKey: Optional[str] = Field(default=None, alias="foreign_key")
    onDelete: Optional[Literal["CASCADE", "SET NULL", "RESTRICT", "NO ACTION"]] = Field(
        default=None, alias="on_delete"
    )
    onUpdate: Optional[Literal["CASCADE", "SET NULL", "RESTRICT", "NO ACTION"]] = Field(
        default=None, alias="on_update"
    )
    
    model_config = {"populate_by_name": True}


# ============================================================================
# INDEX DEFINITION
# ============================================================================

class IndexDefinition(BaseModel):
    """Database index definition"""
    name: str
    fields: list[str]
    unique: Optional[bool] = None


# ============================================================================
# ENTITY DEFINITION
# ============================================================================

class EntityDefinition(BaseModel):
    """
    Domain entity definition
    
    Example:
        entity = EntityDefinition(
            name="Contact",
            description="Customer contact information",
            fields=[
                FieldDefinition(name="id", type="UUID", annotations=FieldAnnotations(generated=True)),
                FieldDefinition(name="email", type="Email", annotations=FieldAnnotations(required=True, unique=True)),
            ],
            relations=[
                RelationDefinition(name="company", type=RelationType.MANY_TO_ONE, target="Company")
            ]
        )
    """
    name: str
    description: Optional[str] = None
    fields: list[FieldDefinition]
    relations: Optional[list[RelationDefinition]] = None
    indexes: Optional[list[IndexDefinition]] = None
    tags: Optional[list[str]] = None


# ============================================================================
# EVENT DEFINITION
# ============================================================================

class EventDefinition(BaseModel):
    """
    Domain event definition
    
    Example:
        event = EventDefinition(
            name="ContactCreated",
            description="Emitted when a new contact is created",
            fields=[
                FieldDefinition(name="contactId", type="UUID", annotations=FieldAnnotations(required=True)),
            ],
            triggers=["api.contact.create", "import.contact"]
        )
    """
    name: str
    description: Optional[str] = None
    fields: list[FieldDefinition]
    triggers: Optional[list[str]] = None
    sourceEntity: Optional[str] = Field(default=None, alias="source_entity")
    
    model_config = {"populate_by_name": True}


# ============================================================================
# WORKFLOW DEFINITION
# ============================================================================

class WorkflowStepType(str, Enum):
    """Workflow step types"""
    ACTION = "action"
    CONDITION = "condition"
    PARALLEL = "parallel"
    LOOP = "loop"
    WAIT = "wait"
    NOTIFY = "notify"
    TRANSFORM = "transform"
    VALIDATE = "validate"


class WorkflowStepDefinition(BaseModel):
    """Workflow step definition"""
    id: str
    type: WorkflowStepType
    name: str
    description: Optional[str] = None
    config: Optional[dict[str, Any]] = None
    next: Optional[str] = None
    onError: Optional[str] = Field(default=None, alias="on_error")
    condition: Optional[str] = None
    
    model_config = {"populate_by_name": True}


class WorkflowDefinition(BaseModel):
    """
    Workflow definition
    
    Example:
        workflow = WorkflowDefinition(
            name="CustomerOnboarding",
            description="New customer onboarding process",
            trigger="ContactCreated",
            steps=[
                WorkflowStepDefinition(id="verify", type=WorkflowStepType.ACTION, name="Verify Email", next="welcome"),
                WorkflowStepDefinition(id="welcome", type=WorkflowStepType.NOTIFY, name="Send Welcome Email"),
            ]
        )
    """
    name: str
    description: Optional[str] = None
    version: Optional[str] = None
    trigger: Optional[str] = None
    schedule: Optional[str] = None
    steps: list[WorkflowStepDefinition]


# ============================================================================
# API DEFINITION
# ============================================================================

ApiOperation = Literal["list", "get", "create", "update", "delete"]


class ApiEndpointDefinition(BaseModel):
    """Custom endpoint definition"""
    method: Literal["GET", "POST", "PUT", "PATCH", "DELETE"]
    path: str
    description: Optional[str] = None
    requestBody: Optional[dict[str, Any]] = Field(default=None, alias="request_body")
    response: Optional[dict[str, Any]] = None
    
    model_config = {"populate_by_name": True}


class ApiResourceDefinition(BaseModel):
    """API resource definition"""
    name: str
    entity: str
    operations: list[ApiOperation]
    customEndpoints: Optional[list[ApiEndpointDefinition]] = Field(
        default=None, alias="custom_endpoints"
    )
    
    model_config = {"populate_by_name": True}


class ApiAuthentication(BaseModel):
    """API authentication configuration"""
    type: Literal["jwt", "api-key", "oauth2", "none"]
    config: Optional[dict[str, Any]] = None


class ApiDefinition(BaseModel):
    """
    API definition
    
    Example:
        api = ApiDefinition(
            version="v1",
            prefix="/api/v1",
            resources=[
                ApiResourceDefinition(name="contacts", entity="Contact", operations=["list", "get", "create", "update", "delete"]),
            ]
        )
    """
    version: str
    prefix: str
    resources: list[ApiResourceDefinition]
    authentication: Optional[ApiAuthentication] = None


# ============================================================================
# APP DEFINITION
# ============================================================================

class AppDefinition(BaseModel):
    """
    Application definition
    
    Example:
        app = AppDefinition(
            name="CRM System",
            version="1.0.0",
            description="Customer Relationship Management System"
        )
    """
    name: str
    version: str
    description: Optional[str] = None
    author: Optional[str] = None
    license: Optional[str] = None


# ============================================================================
# DEFINITION LAYER
# ============================================================================

class DefinitionLayer(BaseModel):
    """
    Layer 1: Definition - defines WHAT to implement
    
    Example:
        definition = DefinitionLayer(
            app=AppDefinition(name="CRM System", version="1.0.0"),
            entities=[
                EntityDefinition(name="Contact", fields=[...]),
            ],
            api=ApiDefinition(version="v1", prefix="/api/v1", resources=[...])
        )
    """
    app: AppDefinition
    entities: list[EntityDefinition]
    events: Optional[list[EventDefinition]] = None
    workflows: Optional[list[WorkflowDefinition]] = None
    api: Optional[ApiDefinition] = None
